### ISSUE TITLE
JIT stopped to work after update 2.1.* to 2.2.2 with laravel mix

### ISSUE THREAD

**Issue by devilstars** [tailwindlabs/tailwindcss#4756]
### What version of Tailwind CSS are you using?

2.2.2

### What build tool (or framework if it abstracts the build tool) are you using?

laravel mix

### What version of Node.js are you using?

v12.9.1

### What browser are you using?

Edge

### What operating system are you using?

Windows

### Reproduction repository

https://github.com/devilstars/opendtf_tall

### Describe your issue

Mix config:
```
const mix = require("laravel-mix");

mix.js("resources/js/app.js", "public/js/app.js")
    .postCss("resources/css/app.css", "public/css", [
        require('tailwindcss')
    ]);

if (mix.inProduction()) {
    mix.version();
}
```

tailwind config
```
const defaultTheme = require('tailwindcss/defaultTheme');

module.exports = {
    mode: 'jit',
    theme: {
        extend: {
            fontFamily: {
                sans: ['Roboto', ...defaultTheme.fontFamily.sans],
            },
        },
    },
    variants: {},
    purge: {
        content: [
            './app/**/*.php',
            './resources/**/*.html',
            './resources/**/*.js',
            './resources/**/*.jsx',
            './resources/**/*.ts',
            './resources/**/*.tsx',
            './resources/**/*.php',
            './resources/**/*.vue',
            './resources/**/*.twig',
        ]
    },
    plugins: [
        require('@tailwindcss/forms'),
        require('@tailwindcss/typography'),
        require('@tailwindcss/aspect-ratio')
    ],
};
```

Expected behavior:
```npm run watch``` - when I update something in my project I get recompiled styles (classes added/removed) every time when I save files and there was something changed

How it works right now:
```npm run watch``` - compiles only once and ignores changes in files

It used to work with tailwind 2.1.4

**Comment by devilstars** [comment]
It seems I've managed to fix it by upgrading:
laravel mix 5 to 6
postcss 7 to 8
post-css loader 3 to 6

Everything seems to work, but incremental rebuilds take about 1 second now. It used to take ~40ms. No clues how to fix it.

**Comment by igorlopasovsky** [comment]
I am having exactly the same issue, unfortunately @devilstars solution doesn't work for me. I am even getting infinite loop on `npm run watch` after upgrade.

**Comment by Lolosan** [comment]
I have infinite loop as well with:

- npx mix watch
- Laravel Mix 6.0.13
- Tailwindcss 2.2.2
- Node 14.17.1 LTS

**Comment by xperator** [comment]
I'm having huge performance issues when enabling JIT mode. The `npm run hot` usually takes around 2seconds to build and update the browser but with tailwind's JIT, it takes around 8seconds which is a LOT and is not something you can live with.

Any ideas?

**Comment by bradlc** [comment]
@igorlopasovsky @Lolosan Please take a look at the troubleshooting docs for help with infinite loops 👍 https://tailwindcss.com/docs/just-in-time-mode#styles-rebuild-in-an-infinite-loop

**Comment by bradlc** [comment]
@devilstars I am not able to reproduce the slowness when saving the CSS file or a template file. Is that what you mean by "incremental rebuild"?

@xperator Are you able to share a repository that reproduces this? Otherwise we won't be able to debug the issue.

**Comment by xperator** [comment]
> @xperator Are you able to share a repository that reproduces this? Otherwise we won't be able to debug the issue.

I'm getting this issue on a working project so can't share that one. But I can provide you with any information/config files you need.
I noticed in the console log, without JIT mode enabled, only the `.vue` I edit will get compiled and updated. But when I enable JIT, it takes it's time to recompile all irrelevant css and sass files such as `bulma.css`, `vendor.css` and `tailwind.css` even though I made no changes to them so that is why it takes extra seconds.
I'm using this in a Laravel + Inertia project.

**Comment by adamwathan** [comment]
@xperator We don't need your actual work project, any project that is setup in a similar way that reproduces the issue is fine. Unfortunately it's not possible to provide good support without reproductions — for you it might take 15 minutes to set up the project for us which will then let us fix the issue quickly, but if we had to spend 30 min or more (since we don't have the full context) trying to create everyone's projects from scratch, we'd never actually have time to fix any issues.

**Comment by xperator** [comment]
@adamwathan Sorry for not replying sooner, but I got so frustrated with the JIT situation not working that I actually spent about 2 days trying to figure out why is it so slow.
I read every discussion about Tailwind JIT and Webpack and I tried every single possible combination of `webpack.mix.js` config. Tried removing many part of my code and turning off many features ( `processCssUrls` , etc...) and also tried different libraries.

 I finally (may have) **found** the culprit. It was this line of code in my `bulma.sass` file:
```sass
@import '../css/fonts.css'
```
When I comment this out, the mix watch/hot recompile takes around 2-3 seconds, instead of 8 ! Which is freaking awesome!

I needed the JIT mode so bad, not only because of the features but also I was hating how tailwind made everything in the browser so laggy and slow, I was having a hard time to work with the inspector and the viewport.

I have two questions though:
1. The Mix/Webpack is still recompiling all the css/sass/tailwind assets everytime I make a change in a vue file. I wonder if that is normal when JIT mode is enabled? Because that is not the case when JIT is disabled.
2. Is the 2-3seconds recompile time a reasonable performance or is it still slow and should be faster? I know tailwind's cli is super fast (avg below 80ms) and I could run that in parallel to mix watch commands ( as suggested in docs) but unfortunately that doesn't integrate with HMR.

**Comment by devilstars** [comment]
> @devilstars I am not able to reproduce the slowness when saving the CSS file or a template file. Is that what you mean by "incremental rebuild"?

"incremental rebuild" is the term used in the official documentation of [tailwind jit](https://tailwindcss.com/docs/just-in-time-mode)

In my case, it should be understood this way: any change in the template files (e.g. adding a class) causes the css file to be updated. This process used to take ~40ms. Now it's more than a second.

The degradation of jit mode performance is about 25 times.

**Comment by adamwathan** [comment]
@devilstars I can't reproduce locally — for me saving a template file triggers a rebuild which takes about ~90ms but this feels like mostly just the overhead of Laravel Mix:

<img width="1247" alt="image" src="https://user-images.githubusercontent.com/4323180/123801211-11413400-d8b8-11eb-808c-919a604a2f48.png">

Switching to Tailwind CLI gives incremental rebuilds closer to ~15ms on average for example:

<img width="1243" alt="image" src="https://user-images.githubusercontent.com/4323180/123801129-f969b000-d8b7-11eb-9c74-48be1bd96c41.png">

@bradlc also tried reproducing using Windows to see if there was a difference there but neither of us can get the project to build as slowly as you're reporting. Any suggestions on what else we can try in order to reproduce?

**Comment by devilstars** [comment]
@adamwathan I've tried to build with Tailwindcss CLI, now it's ~400ms. Not as fast as it used to be, but faster than with Laravel MIX. I'll try to test it with wsl2 later to see if it's windows only issue.

**Comment by adamwathan** [comment]
Got it. You should be able to add the `TAILWIND_MODE=watch` env variable to your scripts and have it work _exactly_ the way it used to in case you want to try that. There are issues with that approach though which is why we've changed the implementation for v2.2. The new more reliable system definitely shouldn't add significant overhead though :/

I'm going to close because I don't think there's anything we can do about this without a reliable way to reproduce, so it's just going to sit here open with no one doing anything about it otherwise. If someone is able to share exact steps to reproduce the slowness reliably, please open a new issue and we will dig into it 👍🏻
